[vcpkg] RFC: Versioning (#11758)
microsoft/vcpkg: docs/specifications/versioning.md

# Versioning Specification 

## 1 Glossary
Some of the terms used in this document have similar o even the same meaning when discussed by the community, and because of that, they can cause confusion and ambiguity. To solve this issue, we will assign specific meaning to these terms and try to keep a consistent usage through the document.

**Library**: A piece of software (source code, binary files, documentation, license, etc.) that is intended to be reutilized by other software.

**Package**: A package can contain a library, collection of libraries, build scripts, software tools, or other components necessary for their use. The goal of VCPKG is to facilitate the installation of these packages in the user's environment.

> **Port**: A VCPKG specific term, a port contains:

Reviewer: Do we have a canonical capitalization of vcpkg? If not we should pick one and use it throughout

Reviewer 2: Personally, I believe that changing between Vcpkg and vcpkg is not the best for having a strong, consistent brand. If there hasn't been an agreed upon capitalization, maybe we could consider keeping it as "vcpkg" or "VCPKG" at all times?

* Metadata about a package: package version, supported features, dependencies, etc.
* Instructions to acquire, build if necessary, and install the package.

## 2 Specifying package versions

> Through the years, C++ software authors have adopted multiple versioning schmes and practices that are sometimes conflicting between each other. On VCPKG, the most recurrent versioning schemes found are:

Reviewer 2: A more concise wording could be "... that sometimes conflict"

> * [Semantic Versioning](semver.org)

Reviewer: Below we say "version-scheme" and the names are singular, so you might want to choose singular instead.

> * [Semantic Versioning](semver.org)

Reviewer 2: ```suggestion
* [Semantic Versioning](https://semver.org/)
```
The previous link leads to a 404 on github

* Dates
* Repository commit IDs
* Arbitrary strings

For VCPKG to achieve wide adoption and interoperability with existing projects it is important that we respect the versioning schemes used by each of the packages contained in our ports library.

Currently, package versions are defined in the Version field of a port's `CONTROL` file. Moving on, we plan to phase out `CONTROL` files in favor of manifest files.

This document describes:

* How to define a package's version, versioning scheme and port revision.
* How to define version requirements for a package's dependencies.
* How VCPKG resolves version requirements.

### 2.1 Manifest files
Package versioning information is specified using the following fields:

**`version-scheme`** 

> The versioning schme used by the package. VCPKG can understand the following values `semver`, `date`, `commit-id`, and `string`. VCPKG uses the field to validate format constraints and determine which version matching operations are supported by the package.

Reviewer: Could this be 'commit' rather than 'commit-id' ?

> * **`semver`**: Accepts version strings that follow semantic versioning conventions. Packages that use this scheme can be sorted which allows VCPKG to apply minimum version requirement constraints on them.

Reviewer: Does this mean that the package in question necessarily follows the semantic version requirements (e.g. bumping the major version when breaking changes happen?) If so, we might want to have an "a.b.c.d" format that doesn't have semver's semantic implications.

>     _Note: some packages that follow semver-like convetions are not actually semver compliant, for examples: `openssl`, and for that reason are more suited to use the arbitrary string versioning scheme._

Reviewer 3: I would argue that we should treat "versions which are like semver", i.e., `([0-9]+)(\.[0-9]+)-.*` should be treated as if they are sortable.

> * **`date`**: Accepts version strings that can be parsed to a date following the format: `YYYY-MM-DD`. Packages that use this scheme can be sorted which allows VCPKG to apply minimum version requirement constraints on them.

Reviewer: ```suggestion
* **`date`**: Accepts version strings that can be parsed to a date following the ISO-8601 format "`YYYY-MM-DD`". Packages that use this scheme can be sorted which allows VCPKG to apply minimum version requirement constraints on them.

```

Author: I really need to study up on my ISO standards.

> * **`date`**: Accepts version strings that can be parsed to a date following the format: `YYYY-MM-DD`. Packages that use this scheme can be sorted which allows VCPKG to apply minimum version requirement constraints on them.

Reviewer: Do we want to allow a suffix / disambiguator in case there is more than one build on a given date? e.g. "2020-05-01.01"

Reviewer 3: I think in general allowing `-.*` on these version strings makes sense.

Author: Yes, I think those are needed.

* **`commit-id`**: Accepts version strings that represent commit IDs under versioning control systems. These are usually hash strings that are not sortable, and as such, VCPKG can only apply direct version requirements on them.

> * **`string`**: Accepts an arbitrary string as the versioning string. The versioning string must be JSON-parsable and may not contain escpaed characters. VCPKG can only apply direct version requirements on packages using this versioning scheme.

Reviewer: All strings are json-parsable so I'm not sure what that requirement means.

> * **`string`**: Accepts an arbitrary string as the versioning string. The versioning string must be JSON-parsable and may not contain escpaed characters. VCPKG can only apply direct version requirements on packages using this versioning scheme.

Reviewer: ```suggestion
* **`string`**: Accepts an arbitrary string as the versioning string. The versioning string must be JSON-parsable and may not contain escaped characters. VCPKG can only apply direct version requirements on packages using this versioning scheme.

```

**`version`**

> The version of the package. The version must match the version of the library or software being installed. VCPKG validates that the string follows the format specified in the `version-scheme` field and reports an error if the version string is not valid.

Reviewer: ```suggestion
The version must match the version of the package being installed. VCPKG validates that the string follows the format specified in the `version-scheme` field and reports an error if the version string is not valid.

```

**`port-revision`**

> Historically, VCPKG port revision numbers were added as a suffix to the version string in `CONTROL` files. This practice must now be disallowed, and port revisions should be instead specified by this field.  Port revisions deal with VCPKG specific changes and do not change the version of the package being installed. 

Reviewer: Is this part of @strega-nil's work with the manifests or part of this?

Reviewer 3: `"port-version"` is how we're specifying this.

> Port revision values should start at 0 for the original revision, increase by 1 each time a new port revision is pu blished, and reset each time the "version" or "version-scheme" fields are changed.

Reviewer: ```suggestion
Port revision values should start at 0 for the original revision, increase by 1 each time a new port revision is published, and reset each time the "version" or "version-scheme" fields are changed.

```

Reviewer 3: Do we have `version-scheme` fields? Are we adding that?

> When resolving package versions only the latest port revision is considered, it is a responsability of the package owner to ensure that no bugs or regressions are introduced when updating the port.

Reviewer: ```suggestion
When resolving package versions, only the latest port revision is considered, it is a responsibility of the port owner to ensure that no bugs or regressions are introduced when updating the port.

```

It can't be the package owner's responsibility to ensure the port is correct since they might not even know the port exists.

Author: Sorry for the typo my Spanish must be leaking (responsabilidad => responsability).

### 2.1.1 Manifest file example

```json
{
    "name": "opencv",
    "version": "4.1.1",

>     "version-scheme": "semver",

Reviewer 3: Should we auto-detect version scheme?

Author: I don't feel like requiring the version-scheme adds too much discomfort. Or at least not enough to warrant the effort of implementing auto-detection.

Reviewer 4: A library may not follow SemVer (e.g. breaking changes in a patch release) but still use version numbers like `"4.1.1"`, which is indistinguishable from a SemVer version number `"4.1.1"`.

    "port-revision": 5
}
```

## 3 Specifying dependency versions

### 3.1 Manifest files
A version requirement can be expressed by using the `"version"` and `"minimum-version"` fields on each dependency. Both are explained in more detail in the [Version Requirements] section.

## 4 CLI changes

### 4.1 Command: `install`
A change to the current install command is needed so users can install specific versions of a package.

The new install syntax follows the pattern:

`vcpkg install package[@version][:triplet]`

For example:

`vcpkg install package zlib@1.2.11:x64-windows`

The version in the `install` command must be an exact match of a package version.

### 4.2 Command: `search`
By default, the search command will only show the latest version of listed packages> To show previous versions of packages. the option `--show-versions` must be passed.

Example:

```
vcpkg search zlib --show-versions
zlib        1.2.11      A compression library
zlib        1.2.10      A compression library
zlib        1.2.8       A compression library
```

> 

Reviewer: We should probably discuss what happens in `upgrade` if:

1. the package version changes
2. the port version changes
3. both change

## 5 Version requirements

### 5.1 Semver version matching
When using semantic versioning is sometimes convenient to shorten the version string when the least significative parts are not relevant. For example, if when talking about a package only the MAJOR and MINOR versions are important, it is convenient to say `2.2` instead of `2.2.0`, `2.2.1`, `2.2.1`, and so on.

In VCPKG, when resolving version requirements, the specificity of the version string is important.

> Truncated parts of a SemVer string are considered as if they were filled with zeroes, e.g.:

Reviewer: I'm not sure I follow, doesn't that mean that a port can't declare that it wants 2.0.1 by saying 2.0 ? It seems like we should interpret missing ones as "latest known". @strega-nil  what do you think?

Reviewer: Edit: below it says we choose 'earliest known'.

Reviewer 4: `2` and `2.1` are incomplete SemVer version numbers and allowing them would add ambiguity - does `2.1` mean `2.1.0` or `2.1.latest` (`~2.1.0` / `>=2.1.0`)? I would expect the latter in a command line context (`vcpkg install foo@2`).

In a manifest, I don't see how it would be convenient to allow a shortened version number. If `2.1` stands for `2.1.0` and then `2.1.1` gets released and you require at least this version, then you have to write out the full version number anyway to define the lower boundary.

* `2` is interpreted as `2.0.0`
* `2.1` is interpreted as `2.1.0`.

### 5.2 No requirements
By not specifying a version requirement, the user implies that the version to install is unimportant. This results in different versions getting installed depending on the circumstances (the composition of the build graph).

> Example: A project depends on package `foo` but doesn't specify a version requirement.

Reviewer: policheck hates 'foo'. How about 'meow'?

* If no version requirements for `foo` are added by a dependency when constructing the build list, the latest version of `foo` is installed.
* If version requirements for `foo` are added by a dependency when constructing the build list:

>   *  And there are no conflicts, the oldest version of `foo` that satisfies all the requirements is installed.

Reviewer: ```suggestion
  * And there are no conflicts, the oldest version of `foo` that satisfies all the requirements is installed.

```

  * And there are conflict, the build fails and the conflicts are reported.

#### 5.2.1 Example

```json
{
    "name": "project",
    "version": "1.0.0", 
    "version-scheme": "semver",
    "dependencies": [ "zlib", "rapidjson" ]
}
```

> ## 5.3 Direct version requirements

Reviewer: Can we rename this to "Exact" rather than "Direct"? Then we don't need additional definition.

The simplest form of versioning. Direct version requirements resolve to an exact version of a package.  
Accomplished by using the `"version"` field in the `"dependencies"` list.

> #### 5.3.1 Example

Reviewer: Is this example a manifest, a lock file, etc.? Where would this example file live?

```json
{
    "name": "project", 
    "version": "1.0.0",
    "version-scheme": "semver",
    "dependencies": [
        { "name": "zlib", "version": "1.2.11" },
        { "name": "rapidjson", "version": "2020-02-08", "version-scheme": "date" }]
}
```

If omitted, `"version-scheme"` defaults to "semver". In the example, `"version-scheme": "date` is required, otherwise an error would be reported as `"2020-02-08"` is not a valid semver string.

## 5.4 Minimum version requirement
A minimum version requirement puts a lower boundary on the versions that can be used to satisfy a dependency. This means that any version that is newer than the requirement is valid (including major version changes).

> VCPKG will use the "oldest" version available that can satisfy all the version requirements in the build graph.

Reviewer: This is a bit scary because it means security patches, which are usually last-octet minor updates, will be dropped on the floor. Example: git 2.16.1 => 2.16.2. If we know that it's semver we might want to choose something like supplied.oldest.latest ?

If not we probably need some discussion as to what port authors are supposed to do when a security patch happens if they want to ensure people upgrade to get the patch.

Reviewer 3: Yeah, this seems unfortunate. This is definitely something I want to talk about in person; can we discuss this in the vcpkg team meeting next week?

Using a minimum version approach has the following advantages:

  * Is predictable and easy to understand.
  * Avoids having to solve version SAT.
  * Avoids unnecesary upgrades when new versions are released.  
  
  _Note: Unexpected downgrades are still possible if a lower version is retro-actively released or added to the package repository. To overcome the issue, the use of Lockfiles is necessary._

Minimum version requirements are expressed by using a "minimum-version" property in the dependencies list. It is not allowed to use the "version" and "minimum-version" properties on the same dependency.

Minimum version requirements can only be used on packages using `semver` or `date` versioning schemes. 

### 5.4.1 Example

```json
{
    "name": "project",
    "version": "1.0.0",
    "dependencies": [
        { "name": "zlib", "minimum-version": "1.2" },

>         { "name": "rapidjson", "minimum-version": "2020-02-01", "version-scheme": "date"}]

Reviewer 3: I really don't like requiring people to specify version schemes here.

}
```

## 6 Design considerations

### 6.1 Problem: version conflicts

#### 6.1.1 Example of a conflict

Consider the following manifest files:

**Project manifest**
```json
{
    "name": "project",
    "version": "1.0",
    "dependencies": [
        { "name": "A", "version": "1.0" },
        { "name": "B", "version": "1.0" }]
}
```

**A's manifest**
```json
{
    "name": "A",
    "version": "1.0.0",
    "dependencies": [{ "name": "C", "version": "1.1" }]
}
```

**B's manifest**
```json
{
    "name": "B",
    "version": "1.0.0",
    "dependencies": [{ "name": "C", "version": "1.2" }]
}
```

There are conflicting requirements for `C`, as package `A` requires version `1.1`, and package `B` requires version `1.2`.

#### 6.1.2 Solution: top-level overrides

> When conflicting requirements exist, an explicit override must be specified in the top-level manifest file.

Reviewer: What is 'top level manifest file'?

The exclusion in example 6.2.1.1:

```json
        { "name": "B", "not": "2.1.0" },
        { "name": "B", "not": "2.1.1" }
```

could be merged into:

```json
        { "name": "B", "not": "2.1.*" }
```

which would exclude all versions that begin with `"2.1"`.

## 7 Algorithm
The implementation is based on the algorithm proposed in https://research.swtch.com/vgo-mvs with some modifications to adapt it to VCPKG version requirements.

To explain the algorithm we will use the following graph to represent the dependencies of a project.

* Blue arrows represent `minimum-version` requirements.
* Green arrows represent exact `version` requirements.

![example package graph image](res/example-package-graph.png)

Also consider the following manifest file for the project:

```json
{
    "name": "project",
    "version": "1.0", 
    "dependencies": [
        { "name": "A", "minimum-version": "1.0" },
        { "name": "B", "minimum-version": "2.0" },
        { "name": "C", "minimum-version": "3.0" }
    ]
}
```

> The original algorithm constrcuts a build list from the dependencies following these steps:

Reviewer: ```suggestion
The original algorithm constructs a build list from the dependencies following these steps:

```

> > Construct the rough build list for M by starting rom an empty list, adding M, and then appending the build list for each of M's requirements. Simplify the rough build list to produce the final build list, by keeping only the newest version of any listed module.

Reviewer: ```suggestion
> Construct the rough build list for M by starting from an empty list, adding M, and then appending the build list for each of M's requirements. Simplify the rough build list to produce the final build list, by keeping only the newest version of any listed module.

```

This says 'keeping the newest version' but above you said 'keeping the oldest version'.

Author: This is the description of the original algorithm. 

But is also something that the proposed algorithm does. When handling `minimum-version`, take the newest `minimum-version` requirement for each package.

For example, if the following requirements are part of the build graph:

* `A >= 1.1` which resolves to `A:1.1` 
* `A >= 1.2` which resolves to `A:1.2`

then we keep the only the newest version `A:1.2` in the list.

Reviewer: I see, if we say 'maximum version **constraint**' or 'maximum minimum version' or something like that I'd have been less confused.

> In VCPKG's case, keeping only the newest versions can result in the violation of direct version requirements. The solution is to create two build lists, one for minimum version requirements (MVR) and one for direct version requirements (DVR). The solutions from these build lists are merged and then either a valid build list is created or a conflict is reported.

Reviewer: I think you could get rid of the abbreviations and just say "minimum version" and "exact version" or "minimum requirement" and "exact requirement" throughout this section to improve clarity.

Reviewer: (for example MVR list => minimum version list)

* Arbitrary string

For VCPKG to achieve wide adoption and interoperability with existing projects, it is important that we respect the versioning schemes used by each of the packages contained in our ports library.

Currently, package versions are defined in the Version field of a port's `CONTROL` file. Moving on, we plan to phase out `CONTROL` files in favor of manifest files.

This document describes:

* How to define a package's version, versioning scheme and port revision.
* How to define version requirements for a package's dependencies.
* How VCPKG resolves version requirements.

### 2.1 Manifest files
Package versioning information is specified using the following fields:

**`version-scheme`** 
The versioning scheme used by the package. VCPKG splits versions into four different kinds: `semver`, `date`, `commit`, and `string`. VCPKG uses the field to validate format constraints and determine which version matching operations are supported by the package.

* **`semver`**: Accepts version strings that follow semantic versioning conventions. Packages that use this scheme can be sorted which allows VCPKG to apply minimum version requirement constraints on them.

>     _Note: some packages that follow semver-like convetions are not actually semver compliant, for examples: `openssl`, and for that reason are more suited to use the arbitrary string versioning scheme._

Reviewer 5: ```suggestion
    _Note: some packages that follow semver-like conventions are not actually semver compliant, for examples: `openssl`, and for that reason are more suited to use the arbitrary string versioning scheme._
```

> * **`date`**: Accepts version strings that can be parsed to a date  following the ISO-8601 format `"YYYY-MM-DD"`. Packages that use this scheme can be sorted which allows VCPKG to apply minimum version requirement constraints on them.

Reviewer 6: ```suggestion
* **`date`**: Accepts version strings that can be parsed to a date following the ISO-8601 format `"YYYY-MM-DD"`. Packages that use this scheme can be sorted which allows VCPKG to apply minimum version requirement constraints on them.

```

* **`commit`**: Accepts version strings that represent commit IDs under versioning control systems. These are usually hash strings that are not sortable, and as such, VCPKG can only apply direct version requirements on them.

> * **`string`**: Accepts an arbitrary string as the versioning string. The versioning string may not contain escaped characters nor any of the following characters: `@`, `:`, `. VCPKG can only apply direct version requirements on packages using this versioning scheme.

Reviewer 4: Are the at symbol, colon and **backtick** reserved?

> * **`string`**: Accepts an arbitrary string as the versioning string. The versioning string may not contain escaped characters nor any of the following characters: `@`, `:`, `. VCPKG can only apply direct version requirements on packages using this versioning scheme.

Reviewer 7: The backtick looks out of place and doesn't match the formatting of the other forbidden characters. This change makes it look the same:
```suggestion
* **`string`**: Accepts an arbitrary string as the versioning string. The versioning string may not contain escaped characters nor any of the following characters: `@`, `:`, `` ` ``. VCPKG can only apply direct version requirements on packages using this versioning scheme.
```

**`version`**
The version must match the version of the package being installed. VCPKG validates that the string follows the format specified in the `version-scheme` field and reports an error if the version string is not valid.

**`port-version`**

> Historically, VCPKG port version numbers were added as a suffix to the version string in `CONTROL` files. This practice must now be disallowed, and port versions should be instead specified by this field.  Port versions deal with VCPKG specific changes and do not change the version of the package being installed. 

Reviewer 6: Why can't it be kept added as a suffix (e.g. like [deb-version](http://manpages.ubuntu.com/manpages/xenial/man5/deb-version.5.html), maybe with a different separating char (e.g. `+`)? Does it complicate the support of the various versioning schemes?
From user point of view it seems easier to look on only one field to understand why a package is upgraded, for example.

##### 6.1.2.1 Example

The example in 6.1.1 can be solved by overriding the requirements of package `C` like this:

```json
{
    "name": "project",
    "version": "1.0",
    "dependencies": [
        { "name": "A", "version": "1.0" },
        { "name": "B", "version": "1.0" }],
    "overrides": [{ "name": "C", "minimum-version": "1.2" }]
}
```

The override is equivalent to changing the version requirements for `C` on both `A`'s manifest and `B`'s manifest to:

```json
{ "name": "C", "minimum-version": "1.2" }
``` 

### 6.2 Problem: block major version upgrades
Under Semantic Versioning, major version changes introduce breaking changes to public API. Using minimum version requirements can result in unexpected major version upgrades of packages in the dependency list. To overcome this issue a user has two options:

> * 1) Override the minimum version requirements

Reviewer 2: ```suggestion
* Override the minimum version requirements
```
Numbering seems a bit redundant with the bullet points already in place

* Exclusions that are not in the top-level manifest are ignored.

An exclusion takes a package version out of the candidate list when resolving version requirements. Effectively, is the same as if the versions in the exclusion list did not exist.

#### 6.2.1.1 Example
Consider that the following package and their versions exist:

* `A: [1.0, 1.1, 2.0]`
* `B: [1.0.0, 2.0.0, 2.1.0, 2.1.1, 2.2.0]`
* `C: [1.0, 1.1, 2.0, 2.1, 3.0]`
* `C:3.0 depends on A:>=2.0`

And the following manifest file

```json
{
    "name": "project",
    "dependencies": [
        { "name": "A", "minimum-version": "1.0" },
        { "name": "B", "minimum-version": "2.0" },
        { "name": "C", "minimum-version": "3.0" }
    ],
    "exclude": [
        { "name": "A", "greater-or-equal": "2.0" },
        { "name": "B", "not": "2.0" },
        { "name": "B", "not": "2.1.0" },
        { "name": "B", "not": "2.1.1" }
    ]
}
```

Without exclusions this would resolve in:

* `A:2.0` (upgraded by `C:3.0`'s requirement).
* `B:2.0.0`
* `C:3.0`

With exclusions the outcome is:

* `A: error` (blocks unexpected upgrade resulting in an error)
* `B: 2.2.0`
* `C: 3.0`

### 6.2.1.2 Wildcard character

> A wildcard character `*` can be used in `not` exclusions for convenience. 

Reviewer 4: NPM's convention is to use `x`, e.g. `2.1.x`
https://semver.npmjs.com/

In our example, to construct a build list for Project we start with its direct requirements, selecting the minimal version that satisfies each requirement. 

* `A >= 1.0` resolves to package `A:1.0`
* `B >= 2.0` resolves to package `B:2.0`. 
* `C >= 3.0` resolves to package `C:3.0`. 

Resulting in: 

* MVR list `[ A:1.0, B:2.0, C:3.0 ]` 
* DVR list `[]`. 

After that we recursively resolve downstream requirements for the packages added to the lists.  

From `A:1.0` 

* `C >= 1.0` resolves to package `C:1.0` 

From `B:2.0` 

* `C:4.0` (direct version requirement) 
* `D:1.0` (direct version requirement) 

From `C:3.0` 

* `E>=1.1` resolves to package `E:1.1` 

From `C:4.0` 

* `E>=1.2` resolves to package `E:1.2` 

Resulting in the temporary lists: 

* MRV list `[ A:1.0, B:2.0, C:1.0, C:3.0, E:1.1, E:1.2]`
* DRV list `[ C:4.0, D:1.0 ]` 

 The lists are then validated resulting in the final temporary lists: 

* MRV list `[ A:1.0, B:2.0, C:3.0, E:1.2 ]` (only latest version of each package)
* DRV list `[ C:4.0, D: 1.0 ]` (no conflicts)

The last step is merging both lists, for which we follow these rules: 

> * If a package appears only in one list, we add the version of that package to the final list. 

Reviewer 2: ```suggestion
* If a package appears only in one list, add the version of that package to the final list. 

```
To keep consistency with the grammar in the following points

To solve this problem a port versions database will now be part of the main repository, the registries spec has more details on how these databases will be implemented in custom registries.  

The form of this database is a folder containing JSON files, one file for each port available, and each file will list all the versions available for a package and how to obtain them from a Git repository.  

As part of the versioning implementation, a generator for these database files will be implemented. The generator will extract from our repository’s Git history, all the versions of each port that had been available at any moment in time and compile them into these database files.  

Example of generated `zlib.json`: 

```json
{ 
  "versions": [ 
    { 
      "git-tree": "2dfc991c739ab9f2605c2ad91a58a7982eb15687", 
      "version": "1.2.11", 
      "port-version": 9 
    }, 
    { “$truncated for brevity” }, 
    { 
      "git-tree": "a516e5ee220c8250f21821077d0e3dd517f02631", 
      "version": "1.2.10", 
      "port-version": 0 
    }, 
    { 
      "git-tree": "3309ec82cd96d752ff890c441cb20ef49b52bf94", 
      "version": "1.2.8", 
      "port-version": 0 
    } 
  ] 
} 
```

> Vcpkg can use the `“git-tree”` objects to acquire (checkout) old versions of ports. 

Reviewer 8: How do you plan to maintain old versions of a port? I didn't find any explanation.

Today `vcpkg` uses one big Git repository to store all packages in the `ports/` directory and tries to use only latest versions. This approach produces a linear versions chain of a port, for example:

`1.0.1 --> 1.0.2 --> 1.0.2#1 --> 1.0.2#2 --> 1.1.1 --> 1.2.0 --> 1.2.4`

With the current ports collection that tends to use latest versions of all ports this is alright but with versioning it isn't. How could you make a fix of a port of a specified old version? I mean something like `1.0.1#1` or `1.0.2#3`? How could you make a port of a lower version when a higher one already exist? For example versions `1.0.3`, `1.1.2` or `1.2.1`. Also versions like 1.1.5 could be released in the upstream after 1.2.0 or 1.2.2 were released, i.e. your may not be able to keep your versions chain in the chronological order. Or you may have some version skipped because of a lack of interest at that time when a newer version was released. For example you may have no 1.5.x version between 1.4.x and 1.6.x but would like to add 1.5.1 after someone already made version 1.6.0 of that port.

With the linear versions chain this is just impossible to achieve. So you need to use Gt **branches** and if you want to use Git branches of port versions you must put each port into a separate Git repository, i.e. make them Git **submodules** of the main `vcpkg` Git repository.

I think the Git workflow of those submodules should be a little bit different to the usual. There should be a `skeleton` branch with only one commit of the port skeleton with upstream version 0 and port version 0. For each real upstream version you should make a new branch and then use those branches to commit future patches, if you need. For example:
```
[skeleton branch]---<skeleton commit>
                       |  |  |
[1.0.1 branch]         |  |  +----<1.0.1 create commit>----<1.0.1#1 fix commit>----<1.0.1#2 fix commit>
[1.0.2 branch]         |  +-------<1.0.2 create commit>----<1.0.2#1 fix commit>
[1.2.0 branch]         +----------<1.2.0 create commit>
```
Each x.y.z branch (or whatever version schema is used by the upstream) should be based on the skeleton branch only. Let's say you want to make `1.0.3` branch for the 1.0.3 upstream version of your port. How to do that? Very simple:

1. `cd` to the Git submodule directory of your port
2. run `git checkout skeleton && git checkout -b 1.0.3`
3. copy files of latest commit in some other branch, for example `git archive 1.0.2 | tar x`
4. modify, add and commit this port files.

With this approach you may use the same JSON files of ports, like your `zlib.json` example, but maintaining of the old port versions will be much easier and more flexible.

Reviewer 9: From what I understand, the problem you describe will be solved [by registries take 2](https://github.com/strega-nil/vcpkg/blob/git-registries/docs/specifications/registries-2.md). This is suppose to be released during this semester along with versioning, as you can see [on the roadmap](https://github.com/microsoft/vcpkg/wiki/Roadmap).

Reviewer 8: @klalumiere That registries design seems over engineered to me. Also it still seems to use a single Git repository for many ports. If a registry is simply a set of packages (as stated in the "take 1" of the registries design document) then it could be just a set of Git repositories where each repository is just for one port/package only. Why to use anything else, for example a filesystem registry with sub-directories for versioning? How could it be maintained without any VCS? Why to invent a new database (the `😇` directory in "take 2") when you can define all submodules (i.e. the registry) in the `.gitmodules` file that Git understand? If some company or organization needs to use a different registry it just needs to use its own version of the `.gitmodules` file. Why to define baselines in that invented database instead of defining proper dependencies with their versions in the ports? The intention of the baseline is to have "a set of versions that are tested against each other". This is the exact purpose of defining proper dependencies with versions.

P.S. There is `git push` after `git commit --amend` in the "take 2" document. How it's supposed to work? You will need to do forced push i.e. `git push -f` but this rewrites history and can make many troubles. I was astonished to see it.

Reviewer 8: One thing I forgot to mention. A Git repository of a ports submodule (of a port) should use Git tags, so all versions of the port could easily be found by `git tag` command, without any other database. For example `zlib-1.2.10`, `zlib-1.2.11-9`, `zlib-1.2.11-11`, i.e. the format is `<port name><first dash><upstream version in any format><optional last dash><optional port version>`.

Reviewer 3: `git push` after `git commit --amend` works if you don't push in between the initial commit and the final commit.

I'm not going to respond to the other questions; if you do want to ask these questions, please ask in the original thread (I will say that likely the things you're asking about do not scale, or require large scale changes across the ecosystem, which are both unacceptable. Existing package managers support on the order of tens of thousands of packages in an automated way, and vcpkg should be able to do the same.).

Reviewer 8: @strega-nil Ok, it's just an unusual practice to split commits, i.e. to use `git commit --amend` for something other than fixing mistakes.

Could you give me a link to that thread?

Reviewer 3: PR #13590 

The reason we have to do this is to grab the tree object ID from git.

Reviewer 10: One thing I have to add here is that the current proposal misses an opportunity for allowing to delegate the maintenance of the portfilesto their respective upstream projects. What @rosti-il suggests with `.gitmodules` would also allow to switch from a vcpkg-handled portfile directly to a vanilla upstream git repos, if the upstream decide to support their portfiles directly. I am sure we all agree that allowing upstream handle their portfiles is something you will eventually have to consider, given the growing popularity of vcpkg and, by extension, the burden of its maintenance.

Reviewer 11: I didn't go through all the conversations above, but I also want to put forward my points. I totally agree with the idea, as @rosti-il mentioned, of using multiple git repositories to maintain port files for each package, and use branches to maintain port files for each version. This scheme has been adopted by many famous package managers like [conda](https://github.com/conda), [boost](https://github.com/boostorg) and [ROS](https://github.com/ros).

By this way, the maintaining burden is further reduced and it enabling modification of port files in previous versions. With current git id versioning, it's impossible to add feature to a previous version of a library (which I mentioned in my issue #2823). That really limited the functionality of vcpkg.

Reviewer 11: Also utilizing git tricks makes vcpkg less user-friendly to developers that are not familiar with git.

{
  "name": "project", 
  "version-semver": "1.0.0",
  "dependencies": [
    { "name": "zlib", "version=": "1.2.11" },
    { "name": "rapidjson", "version=": "2020-02-08.nightly" }
  ]
}
```

## 5.4 Minimum version requirement
A minimum version requirement puts a lower boundary on the versions that can be used to satisfy a dependency. This means that any version that is newer than the requirement is valid (including major version changes).

Vcpkg will use the oldest version available that can satisfy all the version requirements in the build graph.

Using a minimum version approach has the following advantages:

  * Is predictable and easy to understand.
  * User controls when upgrades happen, as in, no upgrades are performed automatically when a new version is released.
  * Avoids having to solve version SAT.  
  
  _Note: Unexpected downgrades are still possible if a lower version is retro-actively released and added to the package registries. To overcome the issue, the use of Lockfiles is necessary._

Minimum version requirements are expressed by using a "version>=" property in the dependencies list. It is not allowed to use "version=" and "version>=" on the same dependency.

> Minimum version requirements can only be used on packages using the `version`, `version-semver` and `version-date` schemes. 

Reviewer 3: Aren't they valid on `version-string` schemas as well because of `port-version`?
